torch/fx/experimental/symbolic_shapes.py

```diff
@@ -4232,6 +4232,17 @@ def evaluate_guards_for_args(self, placeholders, args, *, ignore_static=True):
             return self.evaluate_guards_expression(code, args)
         return True
 
+    def drop_non_input_symint_guards(self, symints):
+        """
+        Removes guards that refer to non input symints
+        """
+        symints = {str(s) for s in symints}
+        guards = []
+        for g in self.guards:
+            if all(str(s) in symints for s in g.expr.free_symbols):
+                guards.append(g)
+        self.guards = guards
```

Comment:
Hmmm it's not clear to me why this is correct. Say we have this code:

```
@torch.compile
def f(x, val):
    if val > 5:
        return x.sin()
    else:
        return x.cos()

x = torch.randn(2)
out1 = f(x, 6)
out2 = f(x, 4)
```

We need to dynamo to fail a guard for `s0 > 5` on the second invocation of `f`. But... if I understand correctly, won't this change in-place mutate the ShapeEnv and delete that guard (while inductor is compiling), before dynamo has a chance to grab all relevant guards from the ShapeEnv and codegen them?

Reply:
@bdhirsh and i looked at this example with dynamic=True and it appears to not fit the cache at all and ends up in a fresh compile